src/cli.cr

```diff
@@ -13,6 +86,163 @@ module DppmRestApi::CLI
+    current_config.users << DppmRestApi::Config::User.new api_key_hash, group_list, required :name
+    current_config.write_to permissions_file
+  end
+
+  def edit_users(match_name, match_groups, api_key, new_name, add_groups, remove_groups, data_dir mn_data_dir, **args)
+    permissions_file = Path[required(:data_dir), "permmissions.json"]
+    current_config = File.open permissions_file do |file|
+      DppmRestApi::Config.from_json file
+    end
+    select_users(match_name, match_groups, api_key, from: current_config.users).each do |user|
+      # edit the users that have made it through all the filters
+      if specified_new_name = new_name
+        user.name = specified_new_name
+      end
+      if groups_to_add = add_groups.try { |arg| split_groups_arg arg }
+        groups_to_add.each { |id| user.join_group id }
+      end
+      if groups_to_remove = remove_groups.try { |arg| split_groups_arg arg }
+        groups_to_remove.each { |id| user.leave_group id }
+      end
+    end
+    current_config.write_to permissions_file
+  end
+
+  def rekey_users(match_name, match_groups, api_key, data_dir mn_data_dir, output_file, **args)
+    permissions_file = Path[required(:data_dir), "permissions.json"]
+    current_config = File.open permissions_file do |file|
+      DppmRestApi::Config.from_json file
+    end
+    using_stdout = true
+    output_io = if o = output_file
+                  using_stdout = false
+                  File.open o
+                else
+                  STDOUT
+                end
+    select_users(match_name, match_groups, api_key, from: current_config.users).each do |user|
+      # Rekey the users that have made it through the filters
+      new_key = Random::Secure.base64 24
+      output_io.puts "user named #{user.name} who has access to the \
+                      groups #{user.groups} is now accessible via API key:"
+      output_io.puts new_key
+      user.api_key_hash = Scrypt::Password.create new_key
+    end
+    output_io.close unless using_stdout
+    current_config.write_to permissions_file
+  end
+
+  def delete_users(match_name, match_groups, api_key, data_dir mn_data_dir, **args)
+    permissions_file = Path[required(:data_dir), "permissions.json"]
+    current_config = File.open permissions_file do |file|
+      DppmRestApi::Config.from_json file
+    end
+    selected = select_users match_name, match_groups, api_key, from: current_config.users
+    current_config.users.reject! { |user| selected.includes? user }
+    current_config.write_to permissions_file
+  end
+
+  def show_users(data_dir, **args)
+    permissions_file = Path[data_dir, "permissions.json"]
+    current_config = File.open permissions_file do |file|
+      DppmRestApi::Config.from_json file
+    end
+    JSON.build STDOUT, indent: 2 do |builder|
+      current_config.users.to_json builder
+    end
+    puts
+  end
+
+  def add_group(id mn_id, name mn_name, permissions mn_permissions, data_dir mn_data_dir, **args)
```

Comment:
It would be simpler to have `add_group(id, name, permissions, data_dir, **args)`, the `String | Nil` -> `String` cast doesn't need an intermediate variable.

Reply:
I don't get what you mean? Of course a cast from nil requires an intermediate variable?

Reply:
No:
```cr
name = if true
  ""
else
  nil
end
puts typeof(name)
name || raise "that's nil!"
puts typeof(name) #=> String
```

Reply:
No:
```cr
name = if true
  ""
else
  nil
end
puts typeof(name) #=> (String | Nil)
name || raise "that's nil!"
puts typeof(name) #=> String
```

Reply:
woaah. I did not know that.

Reply:
woaah. I did not know that.

> Please `raise` instead of `abort`, the exception is then catched with a nice colored `ERR! this is my error` logging.

Ah. :+1: